src/common/XChainIntegrator.sol

```diff
@@ -198,46 +182,124 @@ abstract contract XChainIntegrator is Auth {
 	}
 
 	/*/////////////////////////////////////////////////////
-					Vault Management
+					Address book management
 	/////////////////////////////////////////////////////*/
 
 	function addVault(
 		address _vault,
 		uint16 _chainId,
+		uint16 _postmanId,
 		bool _allowed
-	) external onlyOwner {
-		Vault memory tmpVault = depositedVaults[_vault];
+	) external virtual onlyOwner {
+		_addVault(_vault, _chainId, _postmanId, _allowed);
+	}
 
-		if (tmpVault.chainId != 0 || tmpVault.allowed != false) revert VaultAlreadyAdded();
+	function _addVault(
+		address _vault,
+		uint16 _chainId,
+		uint16 _postmanId,
+		bool _allowed
+	) internal onlyOwner {
+		Vault memory vault = addrBook[_vault];
 
-		depositedVaults[_vault] = Vault(0, _chainId, _allowed);
-		vaultList.push(_vault);
-		emit AddVault(_vault, _chainId);
+		if (vault.chainId != 0) revert VaultAlreadyAdded();
+
+		addrBook[_vault] = Vault(_chainId, _postmanId, _allowed);
+		emit AddedVault(_vault, _chainId);
 	}
 
 	function changeVaultStatus(address _vault, bool _allowed) external onlyOwner {
-		depositedVaults[_vault].allowed = _allowed;
+		addrBook[_vault].allowed = _allowed;
 
-		emit ChangeVaultStatus(_vault, _allowed);
+		emit ChangedVaultStatus(_vault, _allowed);
 	}
 
-	function isVaultAllowed(Message calldata message) external view returns (bool) {
-		return depositedVaults[message.sender].allowed;
+	function updateVaultPostman(address _vault, uint16 _postmanId) external onlyOwner {
+		Vault memory vault = addrBook[_vault];
+
+		if (vault.chainId == 0) revert VaultMissing(_vault);
+
+		addrBook[_vault].postmanId = _postmanId;
+
+		emit UpdatedVaultPostman(_vault, _postmanId);
+	}
+
+	function managePostman(
+		uint16 _postmanId,
+		uint16 _chainId,
+		address _postman
+	) external onlyOwner {
+		postmanAddr[_postmanId][_chainId] = _postman;
+
+		emit PostmanUpdated(_postmanId, _chainId, _postman);
+	}
+
+	/*/////////////////////////////////////////////////////
+					Cross-chain logic
+	/////////////////////////////////////////////////////*/
+
+	function _sendMessage(
+		address receiverAddr,
+		Vault memory vault,
+		Message memory message,
+		messageType msgType
+	) internal {
+		address srcPostman = postmanAddr[vault.postmanId][chainId];
+		address dstPostman = postmanAddr[vault.postmanId][vault.chainId];
+
+		IPostman(srcPostman).deliverMessage(
+			message,
+			receiverAddr,
+			dstPostman,
+			msgType,
+			vault.chainId
+		);
+
+		emit MessageSent(message.value, receiverAddr, vault.chainId, msgType, srcPostman);
+	}
+
+	function receiveMessage(Message calldata _msg, messageType _type) external {
+		// First check if postman is allowed
+		Vault memory vault = addrBook[_msg.sender];
+		if (!vault.allowed || _msg.chainId != vault.chainId) revert SenderNotAllowed(_msg.sender);
+		if (msg.sender != postmanAddr[vault.postmanId][chainId]) revert WrongPostman(msg.sender);
+
+		messageAction[_type](_msg);
```

Comment:
I'm still not convinced that we need to be putting this in storage.
why don't we just add virtual method `_handleMessage` so here we call `_handleMessage(_type, _msg)` and each vault will implement the necessary handling logic

Reply:
Removed this, instead created handleMessage function.